### ISSUE TITLE
dbgeng input, unable to terminate

### ISSUE THREAD

**Issue by jeffbromberger** [NationalSecurityAgency/ghidra#4221]
I'm using Ghidra 10.1.3 and the latest debugging tools from the windows SDK (dbgeng.dll dated 9/11/2021).

One common thing I do in windbg is to edit some memory using the ed command.  It shows the current value, prompts for a new one, and then when you hit enter moves on to the next dword in memory.  When you are done, you simply hit enter without typing new data and that returns back to the normal windbg prompt, i.e. terminates the editing.

But in Ghidra, I am unable to figure out a way to break out of this input phase.  I have tried CTRL-C, CTRL-break, CTRL-Z, the pause icon on the Ghidra GUI to send an interrupt, and, "quit", "exit", of course, hitting enter.  Nothing seems to work.  Is this a bug or am I just missing the magic sequence?  

In the example below I was just trying to update a single dword, so I typed my data for the first location and then did a double-enter.

```
0:000:x86>dd @ecx
014dbb30  0071581c 000000e0 00000001 00000000
014dbb40  00000005 014db2f0 abababab abababab
014dbb50  00000000 00000000 e4746307 0000af81
014dbb60  014c00c0 014cdbe0 feeefeee feeefeee
014dbb70  feeefeee feeefeee feeefeee feeefeee
014dbb80  feeefeee feeefeee feeefeee feeefeee
014dbb90  feeefeee feeefeee feeefeee feeefeee
014dbba0  feeefeee feeefeee feeefeee feeefeee
 0:000:x86>ed @ecx+c
014dbb3c 00000000  >>>1
>>>
>>>dd @ecx
>>>
>>>
>>>
>>>
Interrupt sent
>>>q
>>>exit
>>>
>>>
>>>

```

**Environment (please complete the following information):**
 - OS: Windows 10 21H1 x64
 - Java Version: openjdk 11.0.15 2022-04-19
 - Ghidra Version: 10.1.3
 - Ghidra Origin: official release from github

**Comment by d-millar** [comment]
@jeffbromberger There are some issues with the Interpreter that unfortunately prevent us from making it behave exactly like the windbg console.  (We're working on them, but....)  That said, entering an empty line *should* work, i.e. you should see the prompt switch from ">>>" to "0:000>", and be able to enter commands again.

**Comment by jeffbromberger** [comment]
Ok, let me know if there's anything I can do to help but I'm thinking you should be able to repro easily.  

On a side note, I'm currently trying to figure how to kill the underlying process I was debugging when it got stuck in this input loop.  Since I could never regain control in ghidra I did a disconnect and then closed the debugger.  But the .exe being debugged is still there and since it is marked as being debugged I can't kill it.  Not seeing a way to re-attach to that existing instance in a new invocation of ghidra's debugger or any other way to re-sync this connection.

**Comment by d-millar** [comment]
OK, I think I see the issue.  I started up notepad.exe, and did the following
0:000:x86>dd @rip <return>
00007fff`e60ca2d0  78ec8348 48c98b4c abd2058b 85480010
...
0:000> ed @rip+8  <return>
00007fff`e60ca2d8 abd2058b >>>deadbeef <return>
00007fff`e60ca2dc 85480010 >>> <return>
0:000> dd @rip <return>
00007fff`e60ca2d0  78ec8348 48c98b4c deadbeef 85480010

If I just put in "de", I don't get the prompt back.  Windbg interprets the "de" as the full word value, but our Interpreter does not.  You'd need to either use "eb" or repeat the other bytes in the word.  We can look into this, but I'm going to attribute this to somewhat wanky interaction between the two console flavors. We have a ticket in for an overhaul of the console, but it's going to be complicated and I don't anticipate early resolution on that front.

If you are mid-command and disconnect the debugger, your target is going to be in a weird state.  You'll probably have to kill Ghidra to get the target to die.  If you kill it from the Objects window, it will kill the process, but only after you've completed the command.  I believe you can reconnect to the target via attach if you haven't terminated the session (i.e. disconnected), but once you disconnect all bets are off.  (This is true in windbg too.)  I need to test out the derach/re-attach - am pretty sure I intended to add that functionality, but am also pretty sure I never tested it.

**Comment by jeffbromberger** [comment]
Thanks @d-millar for your follow-up and consideration on this issue.

I tried "eb" as you suggested but I'm not having success.  Two things - I still cannot find a way to stop input when I'm finished, but I also do not get the advance to the next byte to allow me to enter multiple items.  For example, after I type 47 to replace the 46 and hit enter, it should be showing me 4C for the "L" allowing me to change that byte.  At that point if I just hit enter it should quit, or if I give a value it should advance to the next byte.

```
 0:003>da 10eb890
00000000`010eb890  "FLAG{ARBITRARY-READING}........."
00000000`010eb8b0  ""
 0:003>eb 10eb890
00000000`010eb890 46  >>>47
>>>
>>>
>>>
>>>
>>>
>>>0
>>>0x47
>>>
>>>
```

On the other topic of the debugger connection what you describe is perfectly reasonable and not much of a practical limitation.  One small question if I may (sorry to mix two topics here, but the other doesn't seem to be an "issue" so I don't want to create something) - when I first start the debugger on my exe from the launcher it shows a DBGENG@xxxxxx entry in the Debugger Targets window.  But for the life of me I cannot figure out how to actually start debugging that one.  When I go to Debugger -> Debug foo.exe ->in dbgeng locally IN-VM then it fires up the debugger and now I have two entries DBGENG@xxxxxx and DBGENG@yyyyyy.  I'm just unclear on what this first connection is if I'm not actually able to debug with it?

**Comment by d-millar** [comment]
@jeffbromberger Hmmm, something odd is going on, because I'm doing what I believe is exactly the same thing you're doing and getting different results.  That is to say, I start a program, break, do "eb @rip", enter "de", enter "ad", enter "be", and hit one more enter, the prompt returns, and dd returns xxbeadde @rip.  I don't have a good guess yet as to why we're getting different results.

On the launching front, there are several ways to launch a program from the debugger. The most obvious is the bug button in the tool and the associated drop-down.  This will try to run the executable associated with the program in the active tab of the Static listing using the last type of connection you ran, e.g. IN-VM dbgeng.

Alternatively, you can start a debugging session without starting the target and then run/attach/etc. one or more targets from that session.  To do this, you usually use the gear in the Targets window to start the session, and then the bug in the Objects window to launch the executable bound to the program in the Static listing.

You can have multiple sessions up, which sounds like what you did probably inadvertently.

**Comment by jeffbromberger** [comment]
Interesting.  So I started a whole new session of ghidra, relaunched the debugger and used your exact eb @rip sequence and guess what?  It worked :)   Then I did .kill, disconnected the debugger, and clicked the bug button to debug it again.  It attached to the debugger but this time I was seeing some extra error messages about being unable to deliver callbacks.  However, I was still able to toggle my entry point breakpoint and start the execution.  When the bp hits, I tried the exact same eb @rip command and this time it behaves like I was seeing previously.

In all debugger launches I get various errors about not being able to load extension dlls and dbghelp.dll.  I have tried running ghidrarun.bat as admin and this doesn't resolve anything.  Not sure if this is a wow64 issue or what (my exe is 32-bit and the system is 64-bit).  But the sequence seems repeatable - first time debugging the process eb works, then a subsequent time it gets the extra error messages and does not work.  Maybe that is a clue for you?




```
 Unable to add extension DLL:  ntsdexts 
 Unable to add extension DLL:  uext 
 Unable to add extension DLL:  exts 
 SECURE: File not allowed to be loaded -  C:\WINDOWS\SYSTEM32\dbghelp.dll 
Error code:  Win32 error 0n5 
 The call to LoadLibrary( ext ) failed,  Win32 error 0n2 
    " The system cannot find the file specified. "
Please check your debugger configuration and/or network access.
 Unable to add extension DLL:  wow64exts 
 
************* Path validation summary **************
 Response                         Time (ms)     Location
 Deferred                                         srv*C:\symbols*https://msdl.microsoft.com/download/symbols 
 Symbol search path is:  srv*C:\symbols*https://msdl.microsoft.com/download/symbols
 Executable search path is: 
 ModLoad: 00000000`00710000 00000000`0071b000   UserAfterFree.exe
 Unable to deliver callback,  31 
 0:000>g
ModLoad: 00007ff8`d7e30000 00007ff8`d8025000   ntdll.dll
 **Unable to deliver callback,  31** 
 ModLoad: 00000000`774d0000 00000000`77673000   ntdll.dll
 **Unable to deliver callback,  31** 
 20f8:0e68 @ 149747125 - LdrpInitializeProcess - INFO: Beginning execution of UserAfterFree.exe (C:\Users\jeffb\Documents\malware\chat\UserAfterFree.exe)
	Current directory: C:\WINDOWS
	Package directories: (null)
 20f8:0e68 @ 149747125 - LdrpLoadDllInternal - ENTER: DLL name: C:\WINDOWS\SYSTEM32\wow64.dll
 20f8:0e68 @ 149747125 - LdrpFindKnownDll - ENTER: DLL name: wow64.dll
 20f8:0e68 @ 149747125 - LdrpFindKnownDll - RETURN: Status: 0x00000000
 20f8:0e68 @ 149747125 - LdrpMinimalMapModule - ENTER: DLL name: C:\WINDOWS\System32\wow64.dll
 ModLoad: 00007ff8`d6a70000 00007ff8`d6ac9000   C:\WINDOWS\System32\wow64.dll
 **Unable to deliver callback,  31** 

[SNIP]

 20f8:0e68 @ 149753484 - LdrLoadDll - RETURN: Status: 0x00000000
 20f8:0e68 @ 149753484 - LdrpGetProcedureAddress - INFO: Locating procedure "LCIDToLocaleName" by name
 20f8:0e68 @ 149753484 - LdrpGetProcedureAddress - INFO: Locating procedure "LocaleNameToLCID" by name
 Breakpoint 0 hit
 Unable to deliver callback,  31 
 *** WARNING: Unable to verify checksum for  UserAfterFree.exe 
 0:000:x86>db @eip
00713038  e8 2f 05 00 00 e9 4e fe-ff ff 55 8b ec 83 61 04  ./....N...U...a.
00713048  00 83 61 08 00 8b 45 08-89 41 04 8b c1 c7 01 bc  ..a...E..A......
00713058  51 71 00 5d c2 04 00 55-8b ec 56 ff 75 08 8b f1  Qq.]...U..V.u...
00713068  e8 58 00 00 00 c7 06 bc-51 71 00 8b c6 5e 5d c2  .X......Qq...^].
00713078  04 00 83 61 04 00 8b c1-83 61 08 00 c7 41 04 c4  ...a.....a...A..
00713088  51 71 00 c7 01 bc 51 71-00 c3 55 8b ec 56 ff 75  Qq....Qq..U..V.u
00713098  08 8b f1 e8 25 00 00 00-c7 06 d8 51 71 00 8b c6  ....%......Qq...
007130a8  5e 5d c2 04 00 83 61 04-00 8b c1 83 61 08 00 c7  ^]....a.....a...
 0:000:x86>eb @eip
00713038 e8  >>>90
>>>90
>>>
>>>
>>>
>>>

```

**Comment by d-millar** [comment]
Ah, excellent clue - sorry I didn't get back to this today, but will take another stab tomorrow!

**Comment by d-millar** [comment]
@jeffbromberger Well, I wasn't able to reproduce the stop-start thing, but occurs to me I'm not running exactly the same version as you are, i.e. am running in the 10.2 beta.  Your comment about the callback messages made me think of this.  (Am pretty sure that's fixed in 10.2.)  Will try 10.1.3 sometime today - with luck, reproduce the error.  

The "Unable to load" messages probably don't matter.  If you want to get rid of them, copying the dbgXXX.dll's into the bin directory for your JDK will make them go away, I believe, but probably unnecessary.  Running the agent as root and connecting via GADP would also work, but that's an even bigger hassle.  The path sequence is a bit of a nightmare.  We're trying to figure out a better solution, but the fact that Java preloads dbgeng to allow it to do stack walks complicates things.

**Comment by jeffbromberger** [comment]
@d-millar   I'd be happy to try the 10.2 beta on my end if you'd like.  I didn't see it listed under releases, is it available somewhere or not ready yet?

**Comment by d-millar** [comment]
10.2 is equivalent to the current master, i.e. you have to build it from source (which isnt't that hard to do, but....)

**Comment by d-millar** [comment]
OK, wow - so completely reproducible in 10.1.3 and doesn't happen building off of master.  I can spend some time trying to figure out what and how it got fixed, but might be a bit before I get to it.  Are you interested in trying to build Ghidra from source?  That would give you the fix.  :)

**Comment by jeffbromberger** [comment]
@d-millar I would classify my java knowledge as "I've heard of gradle" but nevertheless I did manage to get the thing to build in a WSL2 ubuntu instance.  I then extracted the build/dist zip to my windows setup where I had been running previously (different directory of course).  First thing I noticed was there were no .bat files, only .sh.  I copied the .bats from the 10.1.3 release and it fired up.  But then when I go to disassemble notepad it reports "os/win_x86_64/decompile.exe does not exist".  Doing a recursive dir listing sure enough I don't find that file anywhere in the unzip of what I built, but it is there in the 10.1.3 tree.  Did I need some additional build steps to produce this?

If this is too much of a headache for you I can certainly keep working with 10.1.3 for now.  But if you want me to try out 10.2 I don't mind putting some more time into it...

**Comment by ryanmkurtz** [comment]
When you build Ghidra yourself, it will only build it for the platform you are on.  Since you built in in WSL2, it built it for Linux, which is why you don't have .bat files or win_x86_64 natives files.  You'd either have to run what you built under WSL2, or try the build again on a pure Windows platform (which would require Visual Studio to build the Windows natives).

**Comment by jeffbromberger** [comment]
I got a windows build completed but I'm having trouble launching the debugger now.  It opens my project fine in the code browser, but when I launch the debugger I get an "error in plugin event listener" with this trace

```
java.lang.NullPointerException
	at ghidra.app.plugin.core.debug.gui.objects.components.ObjectTree.getSelectedObject(ObjectTree.java:190)
	at ghidra.app.plugin.core.debug.gui.objects.DebuggerObjectsProvider.getObjectFromContext(DebuggerObjectsProvider.java:322)
	at ghidra.app.plugin.core.debug.gui.objects.DebuggerObjectsProvider.isInstance(DebuggerObjectsProvider.java:866)
	at ghidra.app.plugin.core.debug.gui.objects.DebuggerObjectsProvider.lambda$createActions$19(DebuggerObjectsProvider.java:1027)
	at docking.action.DockingAction.isEnabledForContext(DockingAction.java:184)
	at docking.ComponentPlaceholder.contextChanged(ComponentPlaceholder.java:532)
	at docking.DockingWindowManager.contextChanged(DockingWindowManager.java:2232)
	at docking.AbstractDockingTool.contextChanged(AbstractDockingTool.java:208)
	at ghidra.framework.plugintool.PluginTool.contextChanged(PluginTool.java:1466)
	at docking.ComponentProvider.contextChanged(ComponentProvider.java:450)
	at ghidra.app.plugin.core.debug.gui.objects.DebuggerObjectsProvider.setProgram(DebuggerObjectsProvider.java:466)
	at ghidra.app.plugin.core.debug.gui.objects.DebuggerObjectsPlugin.processEvent(DebuggerObjectsPlugin.java:142)
	at ghidra.framework.plugintool.Plugin.eventSent(Plugin.java:329)
	at ghidra.framework.plugintool.mgr.EventManager.sendEvents(EventManager.java:286)
	at ghidra.framework.plugintool.mgr.EventManager.lambda$new$3(EventManager.java:49)
	at ghidra.util.Swing.doRun(Swing.java:292)
	at ghidra.util.Swing.runNow(Swing.java:208)
	at ghidra.util.Swing.runNow(Swing.java:163)
	at ghidra.framework.plugintool.mgr.EventManager.fireEvent(EventManager.java:216)
	at ghidra.framework.plugintool.PluginTool.firePluginEvent(PluginTool.java:475)
	at ghidra.framework.plugintool.Plugin.firePluginEvent(Plugin.java:481)
	at ghidra.app.plugin.core.progmgr.MultiProgramManager.fireOpenEvents(MultiProgramManager.java:267)
	at ghidra.app.plugin.core.progmgr.MultiProgramManager.addProgram(MultiProgramManager.java:95)
	at ghidra.app.plugin.core.progmgr.ProgramManagerPlugin.lambda$openProgram$6(ProgramManagerPlugin.java:469)
	at ghidra.util.Swing.doRun(Swing.java:292)
	at ghidra.util.Swing.runNow(Swing.java:208)
	at ghidra.util.Swing.runNow(Swing.java:163)
	at ghidra.app.plugin.core.progmgr.ProgramManagerPlugin.openProgram(ProgramManagerPlugin.java:475)
	at ghidra.app.plugin.core.progmgr.ProgramManagerPlugin.openPrograms(ProgramManagerPlugin.java:658)
	at ghidra.app.plugin.core.progmgr.ProgramManagerPlugin.acceptData(ProgramManagerPlugin.java:129)
	at ghidra.framework.plugintool.PluginManager.acceptData(PluginManager.java:48)
	at ghidra.framework.plugintool.PluginTool.acceptDomainFiles(PluginTool.java:414)
	at ghidra.framework.main.ToolButton.openDomainFiles(ToolButton.java:647)
	at ghidra.framework.main.ToolButton$2.end(ToolButton.java:625)
	at ghidra.framework.main.ZoomImageRunner$1.end(ZoomImageRunner.java:54)
	at org.jdesktop.animation.timing.Animator.end(Animator.java:755)
	at org.jdesktop.animation.timing.Animator.stop(Animator.java:660)
	at org.jdesktop.animation.timing.Animator.timingEvent(Animator.java:732)
	at org.jdesktop.animation.timing.Animator.access$200(Animator.java:75)
	at org.jdesktop.animation.timing.Animator$TimerTarget.actionPerformed(Animator.java:1041)
	at java.desktop/javax.swing.Timer.fireActionPerformed(Timer.java:317)
	at java.desktop/javax.swing.Timer$DoPostEvent.run(Timer.java:249)
	at java.desktop/java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:313)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:770)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:721)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:715)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:740)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)

---------------------------------------------------
Build Date: 2022-May-05 0829 CDT
Ghidra Version: 10.2
Java Home: C:\Program Files\Eclipse Adoptium\jdk-11.0.15.10-hotspot
JVM Version: Eclipse Adoptium 11.0.15
OS: Windows 10 10.0 amd64
Workstation: ----
```

**Comment by jeffbromberger** [comment]
Looks like if I proceed on it will still load the debugger.  This does seem to have resolved the input issues I was originally facing.  Even ed is working.  If I type a single byte it'll still advance to the next dword as expected.  And hitting enter breaks out.  So all good there.  Let me know @d-millar if you want me to try anything else related to the above error.  Seems to happen every time I launch the debugger (the Ghidra tool, not a connection to windbg within that tool)
